### DISCUSSION TITLE
Step-by-step addon tutorials

### DISCUSSION THREAD

**Discussion by sfederici** [ScratchAddons/ScratchAddons#6934] [Help]
I would like:

- a step-by-step tutorial on how to build sample ScratchAddons' addons -from very simple to more complex- hopefully explaining how to find the info you need (name/class of the GUI elements you are working on; functions to interact with the Scratch code; etc)
- a step-by-step tutorial on how to adapt a ScratchAddons' addon to Turbowarp -from very simple to more complex- hopefully explaining how to find the info you need (name/class of the GUI elements you are working on; functions to interact with the Scratch code; etc)

Maybe someone of the addons' developers would like to start this tutorial series from their own perspective?

**Comment by CubesterYT** [comment]
This is a great idea, we could have a tutorial series posted to the official YouTube channel!

**Reply by sfederici** [reply]
That would be great

**Comment by WorldLanguages** [comment]
If you're referring to Scratch editor addons, I think it's not that easy, because there aren't any good docs/summaries on the general internals of the Scratch editor, as far as I know. It's pretty hard to create a plugin for a piece of software that isn't well documented, unless you know where to look from prior experience.
In the other hand, creating Scratch website addons is pretty similar to writing browser extensions for any React.js based website.

**Reply by sfederici** [reply]
Yes, thanks for pointing this out, I'm referring to Scratch editor addons indeed. As there aren't any good docs/summaries on the general internals of the Scratch editor, my interest is in making available tutorials made by addons' authors explaining the path they followed to create their addon.

**Reply by WorldLanguages** [reply]
Can you make a list of around 5 to 10 editor addons you would find interesting?

**Reply by sfederici** [reply]
Sure, the following list is what I think it would be interesting. The 10 addons that would be more interesting to me are the ones in bold:
- _how do addons interact with block representation_? 
	- **block-switching**
	- **rename-broadcasts**
	- jump-to-def
	- custom-block-shape
	- custom-block-text
- _how do addons interact with script execution_?
	- **debugger**
	- **pause**
	- default-project
	- editor-stepping
	- load-extensions
- _how to addons interact with sprites thumbnails_?
	- **folders**
	- hide-delete-button
- _how do addons interact with the sprites on the stage_?
	- **disable-stage-drag-select**
- _how do addons interact with Scratch UI_?
	- **editor-compact**
	- **editor-dark-mode**
	- **hide-flyout**
	- hide-stage
	- mediarecorder
	- mouse-pos
	- onion-skinning
	- paint-by-default
	- progress-bar
- _how to addons interact with Scratch menus_?
	- _**rename-broadcasts**_
	- block-switching

**Reply by WorldLanguages** [reply]
Some of the ones in bold aren't that difficult, so we could start with those. The `editor-compact` and `hide-flyout` addons are mostly CSS as far as I know, with some few exceptions, for example the dark background in the costume editor. Possibly `disable-stage-drag-select` would be the most interesting (out of the simple ones) as it involves finding the correct method in the Scratch source code and modifying its behavior.

**Reply by sfederici** [reply]
Excellent

**Reply by mybearworld** [reply]
`scratch-vm` actually [has docs](https://scratchfoundation.github.io/scratch-vm/docs/) and so does [Blockly](https://developers.google.com/blockly/reference/js/blockly) - these power a lot of editor addons

**Reply by WorldLanguages** [reply]
(don't forget Scratch uses a *very old* version of Blockly)

**Reply by mybearworld** [reply]
> (don't forget Scratch uses a _very old_ version of Blockly)

Ohh, that explains a lot

**Reply by ghost** [reply]
hello, i'm fan of scratch addons scratch tools and blocklive. im Linux-Scratcher

**Comment by DNin01** [comment]
I would love to see a more in-depth guide on getting started with developing addons. May as well start with making addon manifests.

**Reply by WorldLanguages** [reply]
Well, that's the boring part, and the one even ChatGPT will understand if you feed it with enough code...
The fun and useful parts of developing consist on interacting with Scratch VM, Blockly and Redux APIs.

**Reply by DNin01** [reply]
> May as well start with making addon manifests.

I guess they meant full tutorials about making certain addons from start to finish, though, so a deep dive into the addon manifest wouldn't be necessary for these, although we would link to the relevant documentation for those interested.

**Comment by sfederici** [comment]
> I guess they meant full tutorials about making certain addons from start to finish, though

Yes, exactly. I would like to learn:
- how addon developers (that, I guess, don't know the full Scratch source code) discover how to interact with the UI and the Scratch internals to get the desired result
- how and when a ScratchAddons' addon is automatically updated (I have seen a pull.js file that generates the _manifest_entry.js and _runtime_entry.js files) and manually updated to work in Turbowarp (I have seen that the js files are different from web and desktop version)

**Reply by WorldLanguages** [reply]
> how and when a ScratchAddons' addon is automatically updated (I have seen a pull.js file that generates the _manifest_entry.js and _runtime_entry.js files)

I'm not sure what you mean by this, what do you exactly mean by "update" here? It's quite simple to load the source code of the extension into your browser and modify any file you want, reload the extension and the page and you're good to go.

**Reply by sfederici** [reply]
I need the addons to work in Turbwarp. In Turbowarp they don't have the classic (I mean: addon.js, userscript.js, userstyle.css) structure. Addon.js is replaced by _manifest_entry.js and _runtime_entry.js. Moreover, the userscript.js has several differences. But I guess I have to post this request to the Turbowarp repository, right?

**Reply by cobaltt7** [reply]
TurboWarp is a completely separate project from SA. They use our addons, yes, but anyone could do that as per the license.

As far as I know, GarboMuffn (the TurboWarp developer) handles TW addon updates on their own, likely using a mix of automation and manual work.

**Reply by sfederici** [reply]
Yes, I know. But as GurboMuffin is in both teams, I keep forgetting it. I'm going to post this last request to the Turbowarp repository.

**Comment by ghost** [comment]
hello, i'love the 3 extention of scratch, i'm Linux-Scratcher
